### DISCUSSION TITLE
MariaDB Replication

### DISCUSSION THREAD

**Discussion by mdance** [ddev/ddev#6130] [Q&A]
I'm attempting to setup mariadb replication, any reason why the server-id is being hardcoded in containers/ddev-dbserver/files/docker-entrypoint.sh:143? Is there a workaround or fix to allow me to change it?  I've tried building a custom docker image but it looks like the dbimage/dbimage_default config flag is gone

**Comment by rfay** [comment]
Where are you replicating from? Where are you replicating to? What is your goal?

You can certainly install your own docker-entrypoint.sh using a custom .ddev/db-build/Dockerfile.* - just change that (https://github.com/ddev/ddev/blob/master/containers/ddev-dbserver/files/docker-entrypoint.sh#L143)

**Comment by mdance** [comment]
I've got a private tailscale network, I do development on my laptop which is running my project in ddev, I'm trying to configure this as the master by adding a .ddev/mysql/replication.cnf as follows:

[mariadb]
log-bin
server_id=1
log-basename=master1
binlog-format=mixed

I've got a remote server with the same ddev project, following a similiar configuration, when I try setting server_id=2, it doesn't take it according to SHOW VARIABLES LIKE 'server_id';

MariaDB [db]> SHOW VARIABLES LIKE 'server_id';
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| server_id     | 1     |
+---------------+-------+
1 row in set (0.002 sec)

My servers .ddev/mysql/replication.cnf is as follows:

[mariadb]
log-bin
server_id=2
log-basename=master1
binlog-format=mixed

When I do a SHOW SLAVE STATUS; its reporting:

Fatal error: The slave I/O thread stops because master and slave have equal MySQL server ids; these ids must be different for replication to work (or the --replicate-same-server-id option must be used on slave but this does not always make sense; please check the manual before using it).

**Comment by mdance** [comment]
Can you clarify where/how exactly I would put the new docker-entrypoint.sh?

I'm looking at https://github.com/ddev/ddev/blob/master/containers/ddev-dbserver/Dockerfile as a reference, and doesn't look like there is anything there I need to override...

https://github.com/ddev/ddev/blob/master/containers/ddev-dbserver/files/docker-entrypoint.sh is the file I'd need to customize, not sure whether I could change line 143 to exec mysqld to just let it read from my replication.cnf, or less preferable exec mysqld --server-id=2

Another oddity is that the line specifies exec mysqld --server-id=0 but the server_id is showing as 1?!?!?!!?

**Reply by rfay** [reply]
The docs are at https://ddev.readthedocs.io/en/stable/users/extend/customizing-images/#adding-extra-dockerfiles-for-webimage-and-dbimage

If I were doing it I'd put the new docker-entrypoint.sh in the .ddev/db-build directory and add a Dockerfile.mariadb that had 

`ADD docker-entrypoint.sh /docker-entrypoint.sh`

**Comment by mdance** [comment]
Thanks for the tips, I actually read that and still didn't think it was possible to do what I'm trying to do

**Reply by rfay** [reply]
It's certainly possible. You just install a new script on top of the existing one. 

But the question remains why you would do it. For local development, you would probably want to implement a hosting provider to do a `ddev pull` to update your local database. https://ddev.readthedocs.io/en/stable/users/providers/

If you're just interested in experimenting with replication, you probably just want to use a regular docker image instead of trying to do it with the DDEV ddev-dbserver.
